source/retryable-writes/retryable-writes.rst

```diff
@@ -181,59 +181,106 @@ Retrying Write Commands
+more than once.
+
+Consider the following pseudo-code:
+
 .. code:: typescript
 
   function executeRetryableWrite(command) {
+    /* Allow ServerSelectionException to propagate to our caller, which can then
+     * assume that no attempts were made. */
     server = selectServer("writable");
 
+    /* UnsupportedException will inform the caller of a configuration error in
+     * their application (i.e. using retryWrites=true with an old cluster) */
     if (server.getMaxWireVersion() < RETRYABLE_WIRE_VERSION) {
       throw new UnsupportedException();
     }
 
+    /* NetworkException and NotMasterException are both retryable errors. If
+     * caught, remember the exception, update SDAM accordingly, and proceed with
+     * retrying the operation. */
     try {
       return executeCommand(server, command);
-    } catch (NetworkException e) {
-      updateTopologyDescriptionForNetworkError(server, e);
-    } catch (NotMasterException e) {
-      updateTopologyDescriptionForNotMasterError(server, e);
+    } catch (NetworkException originalError) {
+      updateTopologyDescriptionForNetworkError(server, originalError);
+    } catch (NotMasterException originalError) {
+      updateTopologyDescriptionForNotMasterError(server, originalError);
     }
 
-    server = selectServer("writable");
+    /* If we cannot select a writable server, do not proceed with retrying and
+     * throw the original error. The caller can then infer that an attempt was
+     * made and failed. */
+    try {
+      server = selectServer("writable");
+    } catch (Exception ignoredError) {
+      throw originalError;
+    }
 
-    // If the new server is too old, throw original error
+    /* If the server selected for retrying is too old, throw the original error.
+     * The caller can then infer that an attempt was made and failed. This case
+     * is very rare, and likely means that the cluster has mixed versions, just
+     * experienced a fail over, and we missed the error that retryable writes
+     * were not supported during the first attempt (3.6 feature compatibility
+     * could not have been enabled). */
     if (server.getMaxWireVersion() < RETRYABLE_WIRE_VERSION) {
-      throw e;
+      throw originalError;
     }
 
-    return executeCommand(server, command);
+    /* Allow any retryable error from the second attempt to propagate to our
+     * caller, as it will be just as relevant (if not more relevant) than the
+     * original error. For any other exception, we should raise the original
+     * error. */
+    try {
+      return executeCommand(server, command);
+    } catch (NetworkException secondError) {
+      updateTopologyDescriptionForNetworkError(server, secondError);
+      throw secondError;
+    } catch (NotMasterException secondError) {
+      updateTopologyDescriptionForNotMasterError(server, secondError);
+      throw secondError;
+    } catch (Exception ignoredError) {
+      throw originalError;
```

Comment:
Does this part mean that a network error followed by a DuplicateKeyError will result in the original network error being raised?

Reply:
That wasn't my intention. According to [Bulk Update: On Errors](https://github.com/mongodb/specifications/blob/master/source/driver-bulk-update.rst#on-errors), the command still returns `ok: 1` when it includes write and write concern errors. I would not expect `executeCommand()` in the preceding pseudo code to throw in this case and assume that any logic that creates a BulkWriteException (or other write exception) from the `writeErrors` and `writeConcernError` fields in a merged write result object will happen in some parent context outside of `executeRetryableWrite()`.

Reply:
So which exceptions do you intend to catch here?

Reply:
This was added in response to @jyemin's comment in https://github.com/mongodb/specifications/pull/228#discussion_r146321727. I'll note that this psuedo-code deviates from the preceding prose text, which still says:

> If the retry attempt also fails, drivers MUST update their topology according to the SDAM spec (see: Error Handling) and raise any ensuing error to the user. In this case, the error from the retry attempt is just as relevant as the original retryable error and the caller is still able to infer that an attempt was made.

Any suggestions for how best to clarify this? It seems somewhat related to "hard stop" errors I talk about in [SPEC-976](https://jira.mongodb.org/projects/SPEC/issues/SPEC-976).

Reply:
Thanks for the context. Maybe `catch (DriverException ignoredError)` and explain what a `DriverException` could be?

Reply:
https://github.com/mongodb/specifications/pull/229